### PR TITLE
[POAE7-2730] Unary StringOp support in Nextgen (upper/lower/char_length)

### PR DESCRIPTION
**Pull request by YBRua**
### What changes were proposed in this pull request?

Added support for unary StringOp scalar functions in nextgen (including UPPER/LOWER/CHAR_LENGTH), along with a few bug fixes in StringOp workflow.

- StringOp Support
  - Added support for case conversion (UPPER/LOWER). Implementation of related runtime functions was based on DuckDb's implementataion.
  - Added support for computing char length (CHAR_LENGTH).
  - Enabled related test cases.
- Fixes
  - Fixed instantiation of `StringOp` base class, which breaks virtual function inheritance.
  - Fixed pointers in StringOp args not being correctly updated (because a copy of the pointers where returned).

### Why are the changes needed?

Nextgen migration.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

UTs.

### CODE REVIEW

File: src/cider/exec/nextgen/function/CiderStringFunctions.cpp [hunk 1]
```diff
@@ -76,3 +76,64 @@ extern "C" RUNTIME_EXPORT int32_t format_substring_len(int pos,
...
+extern "C" RUNTIME_EXPORT int64_t cider_ascii_lower(int8_t* string_heap_ptr,
+                                                    const char* str,
+                                                    int str_len) {
+  StringHeap* ptr = reinterpret_cast<StringHeap*>(string_heap_ptr);
+  string_t s = ptr->emptyString(str_len);
+  char* sout = s.getDataWriteable();
+  for (int i = 0; i < str_len; ++i) {
+    sout[i] = ascii_char_lower_map[str[i]];
+  }
+  return pack_string((const int8_t*)s.getDataUnsafe(), (const int32_t)s.getSize());
```
Comment (jikunshang) [inline src/cider/exec/nextgen/function/CiderStringFunctions.cpp:126]: can create a new method like `int64_t pack_string_t(string_t s)`

File: src/cider/exec/nextgen/function/CiderStringFunctions.cpp [hunk 2]
```diff
@@ -76,3 +76,64 @@ extern "C" RUNTIME_EXPORT int32_t format_substring_len(int pos,
...
+    238, 239, 240, 241, 242, 243, 244, 245, 246, 247, 248, 249, 250, 251, 252, 253, 254};
+
+extern "C" RUNTIME_EXPORT int64_t cider_ascii_lower(int8_t* string_heap_ptr,
+                                                    const char* str,
+                                                    int str_len) {
+  StringHeap* ptr = reinterpret_cast<StringHeap*>(string_heap_ptr);
+  string_t s = ptr->emptyString(str_len);
+  char* sout = s.getDataWriteable();
+  for (int i = 0; i < str_len; ++i) {
+    sout[i] = ascii_char_lower_map[str[i]];
```
Comment (jikunshang) [inline src/cider/exec/nextgen/function/CiderStringFunctions.cpp:124]: `sout[i] = ascii_char_lower_map[(uint8_t*)str[i]];` would be better?

File: src/cider/exec/nextgen/function/CiderStringFunctions.cpp [hunk 3]
```diff
@@ -26,6 +26,11 @@ ALWAYS_INLINE uint64_t pack_string(const int8_t* ptr, const int32_t len) {
          (static_cast<const uint64_t>(len) << 48);
 }
 
+ALWAYS_INLINE uint64_t pack_string_t(const string_t& s) {
+  return (reinterpret_cast<const uint64_t>(s.getDataUnsafe()) & 0xffffffffffff) |
```
Comment (jikunshang) [inline src/cider/exec/nextgen/function/CiderStringFunctions.cpp:30]: call `pack_string` instead

File: src/cider/type/plan/StringOpExpr.cpp [hunk 4]
```diff
@@ -348,14 +344,115 @@ JITExprValue& SubstringStringOper::codegen(JITFunction& func) {
...
+  auto ret_len = func.emitRuntimeFunctionCall(
+      "extract_string_len",
+      JITFunctionEmitDescriptor{.ret_type = JITTypeTag::INT32,
+                                .params_vector = {ptr_and_len.get()}});
+
+  return set_expr_value(arg_val.getNull(), ret_len, ret_ptr);
+}
+
+// UpperStringOper: UPPER
+
```
Comment (jikunshang) [inline src/cider/type/plan/StringOpExpr.cpp:399]: remove empty line here

File: src/cider/type/plan/StringOpExpr.cpp [hunk 5]
```diff
@@ -348,14 +344,115 @@ JITExprValue& SubstringStringOper::codegen(JITFunction& func) {
...
-      "extract_str_len",
+      "extract_string_len",
       JITFunctionEmitDescriptor{.ret_type = JITTypeTag::INT32,
                                 .params_vector = {ptr_and_len.get()}});
+
   return set_expr_value(arg_val.getNull(), ret_len, ret_ptr);
 }
 
+// CharLengthStringOp: CHAR_LENGTH
+
```
Comment (jikunshang) [inline src/cider/type/plan/StringOpExpr.cpp:440]: remove empty line here

**Review by jikunshang — DISMISSED** [review]

**Review by jikunshang — APPROVED** [review]

File: src/cider/exec/template/ExpressionRewrite.cpp [hunk 6]
```diff
@@ -735,11 +739,16 @@ class ConstantFoldingVisitor : public DeepCopyVisitor {
       // However, Cider does not handle chained_string_op_exprs_
       // so we should return the original stringop here
       // or otherwise the folded stringops will be mistakenly dropped and never processed
-      return makeExpr<Analyzer::StringOper>(kind, return_ti, rewritten_args);
+      // return makeExpr<Analyzer::StringOper>(kind, return_ti, rewritten_args);
```
Comment (yma11) [inline src/cider/exec/template/ExpressionRewrite.cpp:742]: This line comment should be removed.

File: src/cider/exec/template/ExpressionRewrite.cpp [hunk 7]
```diff
@@ -735,11 +739,16 @@ class ConstantFoldingVisitor : public DeepCopyVisitor {
...
+      // NOTE: (YBRua) do not instantiate StringOper base class, it breaks virtual
+      // functions such as codegen() in nextgen framework.
+      // construct derived classes instead
+      return makeStringOperExpr(kind, return_ti, rewritten_args);
     } else {
       CHECK(!in_string_op_chain_);
-      return makeExpr<Analyzer::StringOper>(
-          kind, return_ti, rewritten_args, chained_string_op_exprs_);
+      // return makeExpr<Analyzer::StringOper>(
+      //     kind, return_ti, rewritten_args, chained_string_op_exprs_);
```
Comment (yma11) [inline src/cider/exec/template/ExpressionRewrite.cpp:750]: remove above 2 lines directly.

File: src/cider/tests/nextgen/functional/CiderStringNextgenTest.cpp [hunk 8]
```diff
@@ -332,8 +332,6 @@ TEST_F(CiderStringRandomTestNextGen, NestedSubstringTest) {
 // stringop: upper/lower
```
Comment (yma11) [inline src/cider/tests/nextgen/functional/CiderStringNextgenTest.cpp:332]: Shorten this test file name to `StringTest`?

**Review by yma11 — APPROVED** [review]

**Review by jikunshang — APPROVED** [review]

**Pull request merged**